### ISSUE TITLE
[Feature] Support shorter dates like YYYY-MM and YYYY

### ISSUE THREAD

**Issue by Telkhine** [stashapp/stash#1576]
The day of month or even the month is not known on many files. Currently only YYYY-MM-DD date format is supported. Could we also support YYYY-MM and YYYY?

**Is your feature request related to a problem? Please describe.**
When I try to enter a YYYY-MM or YYYY as the date, I get no error. It seems like it's going to accept the value, but it doesn't get saved.

**Describe the solution you'd like**
I'd like see support for YYYY-MM and YYYY, e.g. 2021 or 2021-07.

**Describe alternatives you've considered**
If for some reason the month and day of month can't be blank, maybe allow for something like this: 2021-00-00

**Comment by gitgiggety** [comment]
While IMO a valid request this sadly isn't possible. This as most, if not all, date fields are stored in the database as an actual date (this so it's also possible to make calculations, like performers age or performers age in scene). But this at the same time thus doesn't allow to enter invalid dates. So the only work around would be to enter a valid date like 2021-01-01.

**Comment by Telkhine** [comment]
That's what I currently do, but it's very messy and misleading. Probably more than half the files I have, I don't know the day or month for.

How about this:
If a user only enters YYYY-MM or YYYY, it auto translates the day/month to 01, then hides it from the user.

**Ex**: 2011 becomes 2011-01-01 in the DB, but on the frontend it only shows 2011.

If a user acually enters 2011-01-01, it would obvioulsy show the full date.

**Comment by Telkhine** [comment]
I've been thinking about this. The other way the could be implemented is if dates were stored as integers, then cast to a date format when needed. If only the year was stored, then a casting function would automatically assume January 1 as the month and day. There would also need to be valid date checking before a value is stored.

**Comment by DogmaDragon** [comment]
> While IMO a valid request this sadly isn't possible. This as most, if not all, date fields are stored in the database as an actual date (this so it's also possible to make calculations, like performers age or performers age in scene). But this at the same time thus doesn't allow to enter invalid dates. So the only work around would be to enter a valid date like 2021-01-01.

The backend allows storing partial date, it's the UI that doesn't allow it.

**Comment by GrumpyRussian** [comment]
Please note that this feature will allow for preparation of submissions to StashDB with partial dates right in StashApp.

**Comment by BonerFide** [comment]
My thought on how to do this actually would be to store an exact date (or not) and also a ‘precision’ field. I have had the opposite issue where in stash I would occasionally benefit from both the time as well (eg something like cam recording’s currently don’t order chronologically without special naming).

We would also benefit from UTC offset support in both stash and stash DB for some content (eg often we get ‘off by one’ dates in submissions depending on where the submitter is, in say New Zealand or Hawaii.

So yeah being able to say this date is certain to the nearest minute/day/month/year etc would be useful. Also for performers these days for privacy you may only know the birthdate with +/- 1 year of precision.

**Comment by gitgasm** [comment]
One use case where a Year `YYYY` field available is the Plex use case. Since Plex expects files to have the following filenames `Title (YYYY).ext` for it to match automatically, being able to define title and year in the RenameFile or RenamerOnUpdate plugins would help this process work smoothly.

**Comment by genetics-mantra** [comment]
I totally feel the request. From my point of view there should also be a possibility to only enter the Month and Day (MM-DD). In my experience many times you find the birthday of performers, but they do not publish the year/age.

**Comment by bbappserver** [comment]
I have a personal tool that encodes partial dates using a separate enum. this works just fine with sqlite
https://github.com/bbappserver/retcon/blob/2e6f3ce364e69ec0d92b3a9896cc5b8a4e27603c/backend/retconcreatives/admin.py#L87-L94

It uses a date parsing library and automatically fills the precision field for the user if able. So I have a working reference implementation with a sqlite db, though of course Django is Python not Go.

<img width="443" height="466" alt="Image" src="https://github.com/user-attachments/assets/e9607152-fd52-46ae-bd68-cfe36f973e01" />

**Comment by Gykes** [comment]
So this may be a larger change than I initially thought unless I am wrong.

Currently it looks like such:
- User types "2025" into the date field
-  ParseDateStringAsTime passes 2025 to time.time
- time.time converts to full date format (YYYY-MM-DD)
-  Stored in the DB and returned to frontend as such

To do this I believe we will need to do something similar to what stashbd did to allow dates to be stored in the DB as straight strings. The problem with changing to storing as straight strings is I believe this will break things such as comparing. Currently we use time.Time to convert INTO the database and we could potentially remove that and then convert on it's way out as needed. This will allow users to add in partial dates but we would need to refactor everything that currently requires a fully formatted date to convert OUT of the db.

I have it mocked up currently to where a user can type in the field 2025 and it saves as 2025-01-01 not worth PRing in my opinion.

**Comment by WithoutPants** [comment]
#6305 added support for accepting partial dates in the backend; setting the unpassed components as `01`. Since we can do this in the backend, it makes sense to support it in the frontend. Doing so won't complete this feature request, but it will improve things for people that don't mind if the unset parts are `01` and is a good first step.

Adding full support for partial dates is a larger change, and wasn't something that I suggested for 0.30 scope.

**Comment by Gykes** [comment]
`I have it mocked up currently to where a user can type in the field 2025 and it saves as 2025-01-01 not worth PRing in my opinion.`

You essentially just want this then?

**Comment by WithoutPants** [comment]
Yep. The placeholder text should probably also be updated, but I'm not sure of a clean way to show it.

**Comment by sleetx** [comment]
> [#6305](https://github.com/stashapp/stash/pull/6305) added support for accepting partial dates in the backend; setting the unpassed components as `01`. 
> it will improve things for people that don't mind if the unset parts are `01` and is a good first step.

I'm not a fan of this approach. Any Jan 1st date moving forward will become suspect even if it is legitimate.

There are several other approaches people have mentioned in this thread that do not involve setting a fully valid date in situations where the date is partially unknown.

**Comment by RyanCrossby** [comment]
I agree with @sleetx, I find this change backwards, as it can be misleading, and there is no way to differentiate from real 01s. As mentioned above, even a temporary precision field (stating if the date is complete / month-year only / year only) would be appreciated, so that we can know what to migrate once a more nuanced solution is found.

**Comment by WithoutPants** [comment]
I'm going to look into the possibility of storing date precision in existing date fields, with the intention to include in 0.30.

I'm anticipating the implementation being to add a precision field to each date field in the database. The presentation will return date strings to the applicable precision, and accept `YYYY-MM-DD`, `YYYY-MM` or `YYYY`, setting the precision accordingly.

**Comment by BonerFide** [comment]
> I'm anticipating the implementation being to add a precision field to each date field in the database. The presentation will return date strings to the applicable precision, and accept `YYYY-MM-DD`, `YYYY-MM` or `YYYY`, setting the precision accordingly.

While we're in here changing this can we also allow minute or second precision as options as well? This would be really useful for both cam recordings and images that often have multiple entries per day and really benefit from being in a more precise timestamp 'date' order especially when not naturally that way via name order. 

I imagine date as just being a very exact date/time field (eg ISO8601 string - YYYY-MM-DD HH:MM:SS.SSS / or unix timestamp) and the precision just being set in the DB for UX purposes to show how precise we 'know' this field is and for display.

Personally I would avoid using strings though except for display. The separate precision field allows sorting to work really well even if for sorting purposes it's assuming that an absent day/month is the 1st and an absent time is 00:00:00. If stashbox is storing 2025 as "2025" instead of "2025-01-01 00:00:00.000" and a precision of 1 (for example) that's a code smell IMO, given the performance implications on a sort.

**Comment by WithoutPants** [comment]
> While we're in here changing this can we also allow minute or second precision as options as well? This would be really useful for both cam recordings and images that often have multiple entries per day and really benefit from being in a more precise timestamp 'date' order especially when not naturally that way via name order.

This is out of scope for this change. Please raise this as a separate feature request.
